### PR TITLE
[14.0][IMP] base_revision: Add revisions_count field + update the state field if it exists in the model.

### PR DESCRIPTION
**Pull request by victoralmau**
Add revisions_count field + update the state field if it exists in the model.

Please @pedrobaeza and @CarlosRoca13 can you review it?

@Tecnativa

### CODE REVIEW

File: base_revision/models/base_revision.py [hunk 1]
```diff
@@ -38,6 +38,20 @@ def _compute_has_old_revisions(self):
     )
     active = fields.Boolean(default=True)
     has_old_revisions = fields.Boolean(compute="_compute_has_old_revisions")
+    revisions_count = fields.Integer(compute="_compute_revisions_count")
```
Comment (pedrobaeza) [inline base_revision/models/base_revision.py]: Don't use plural:
```suggestion
    revision_count = fields.Integer(compute="_compute_revisions_count")
```

File: base_revision/models/base_revision.py [hunk 2]
```diff
@@ -74,9 +88,10 @@ def copy_revision_with_context(self):
         default_data.update(vals)
         new_revision = self.copy(default_data)
         self.old_revision_ids.write({"current_revision_id": new_revision.id})
-        self.write(
-            {"active": False, "state": "cancel", "current_revision_id": new_revision.id}
-        )
+        vals = {"active": False, "current_revision_id": new_revision.id}
```
Comment (pedrobaeza) [inline base_revision/models/base_revision.py]: Add a `_prepare` hook method for this for easing possible extension

File: base_revision/models/base_revision.py [hunk 3]
```diff
@@ -38,6 +38,20 @@ def _compute_has_old_revisions(self):
...
+        res = self.with_context(active_test=False).read_group(
+            domain=[("current_revision_id", "in", self.ids)],
+            fields=["current_revision_id"],
+            groupby=["current_revision_id"],
+        )
+        revision_dict = {
+            x["current_revision_id"][0]: x["current_revision_id_count"] for x in res
+        }
+        for rec in self:
+            rec.revision_count = revision_dict.get(rec.id, 0)
```
Comment (pedrobaeza) [inline base_revision/models/base_revision.py:56]: Shouldn't be + 1 for the current revision + old revisions?

File: base_revision/models/base_revision.py [hunk 4]
```diff
@@ -67,16 +81,20 @@ def _get_new_rev_data(self, new_rev_number):
             "old_revision_ids": [(4, self.id, False)],
         }
 
+    def _prepare_revision_data(self, new_revision):
+        vals = {"active": False, "current_revision_id": new_revision.id}
+        if "state" in self._fields:
+            vals.update({"state": "cancel"})
```
Comment (pedrobaeza) [inline base_revision/models/base_revision.py]: Why assume that the inherited model is going to have such selection value in `state`? Do this in the inherited model. That's why we have put the hook.

**Review by dreispt — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Aren't you labeling fields with the texts we talked?

**Comment by pedrobaeza** [comment]
We talked about putting the smart button as "Prev. revisions", and the fields as "Previous versions count".

**Comment by victoralmau** [comment]
> We talked about putting the smart button as "Prev. revisions", and the fields as "Previous versions count".

I have defined "Previous versions count" in the field, but the smart-button will be added here https://github.com/OCA/sale-workflow/pull/1911 when this PR is merged.

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**